### PR TITLE
[SPARK-23340][SQL] Upgrade Apache ORC to 1.4.3

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR updates Apache ORC dependencies to 1.4.3 released on February 9th. Apache ORC 1.4.2 release removes unnecessary dependencies and 1.4.3 has 5 more patches (https://s.apache.org/Fll8).

Especially, the following ORC-285 is fixed at 1.4.3.

```scala
scala> val df = Seq(Array.empty[Float]).toDF()

scala> df.write.format("orc").save("/tmp/floatarray")

scala> spark.read.orc("/tmp/floatarray")
res1: org.apache.spark.sql.DataFrame = [value: array<float>]

scala> spark.read.orc("/tmp/floatarray").show()
18/02/12 22:09:10 ERROR Executor: Exception in task 0.0 in stage 1.0 (TID 1)
java.io.IOException: Error reading file: file:/tmp/floatarray/part-00000-9c0b461b-4df1-4c23-aac1-3e4f349ac7d6-c000.snappy.orc
	at org.apache.orc.impl.RecordReaderImpl.nextBatch(RecordReaderImpl.java:1191)
	at org.apache.orc.mapreduce.OrcMapreduceRecordReader.ensureBatch(OrcMapreduceRecordReader.java:78)
...
Caused by: java.io.EOFException: Read past EOF for compressed stream Stream for column 2 kind DATA position: 0 length: 0 range: 0 offset: 0 limit: 0
```

## How was this patch tested?

Pass the Jenkins test.

### CODE REVIEW

**Comment by kiszk** [comment]
LGTM since 1.4.2 does not seem to include major changes based on [the release summary](https://orc.apache.org/news/2018/01/23/ORC-1.4.2/).

**Comment by dongjoon-hyun** [comment]
Right, @kiszk !

**Comment by dongjoon-hyun** [comment]
The failure is due to flaky test suite.
```
org.apache.spark.sql.hive.client.HiveClientSuites.(It is not a test it is a sbt.testing.NestedSuiteSelector)
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by gatorsmile** [comment]
Just FYI, this will not be merged to 2.3

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you for review, @gatorsmile and @HyukjinKwon .
Sure, this is for Apache Spark 2.4.

**Comment by HyukjinKwon** [comment]
cc @srowen

**Review by felixcheung — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by kiszk** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you, @kiszk .

**Comment by gatorsmile** [comment]
I am wondering what is the difference between ORC 1.4.3 and ORC 1.4.2? Their release notes are the SAME.  https://orc.apache.org/news/2018/02/09/ORC-1.4.3/ and https://orc.apache.org/news/2018/02/09/ORC-1.4.2/ Could you help us figure out the exact change JIRA lists excluded in these two releases? 

Should we directly upgrade it to 1.4.3? What is the release schedule for Apache ORC? Our Spark 2.4 will not be released until the second half of 2018. Which version of ORC is stable for production? I am wondering if we should always upgrade to the latest version of ORC? Or wait for more user feedbacks from the ORC community to know whether the version is stable or not?

cc @dongjoon-hyun @omalley

**Comment by omalley** [comment]
Sorry, I forgot to transition the jira issues for the ORC 1.4.3, so they didn't show up in the search from the notes.

The list of jiras closed by the 1.4.3 release is: https://s.apache.org/Fll8

There was an issue with the reader if you had an empty column of floats/doubles (ORC-285) and a compression issue that only seemed to hit LLAP (ORC-296).

We are about to start the ORC 1.5 release, but the ORC 1.4 release has been very stable.

**Comment by gatorsmile** [comment]
@omalley Thanks for your quick reply!

@dongjoon-hyun Maybe we should directly bump to 1.4.3?

**Comment by kiszk** [comment]
+1 for 1.4.3

**Comment by dongjoon-hyun** [comment]
Sure. No problem. BTW, is it applicable for Apache Spark 2.3?

**Comment by gatorsmile** [comment]
@dongjoon-hyun Could you go over the list of the resolved JIRAs in ORC 1.4.2 and 1.4.3 that could cause the regressions?

We need to know the impact and the risk. If possible, also added a test case in Spark to ensure the issue has been resolved after the upgrade?

**Comment by dongjoon-hyun** [comment]
First of all, ORC 1.4.2 was very safe because it has only ORC-235 removing redundant dependencies.

For ORC 1.4.3, the following five patches are included. 

1. ORC-298 Move the benchmark code base to non-Apache repository
2. ORC-240 Fix warnings from Maven
3. ORC-217 Duplicate rat plugins in pom.xml

The above three are trivial.

4. ORC-285 Empty vector batches of floats or doubles get  java.io.EOFException
5. ORC-296 Work around HADOOP-15171; also fix stream contract

(4) is only adding a workaround for `batchSize=0`.  (5) may cause performance difference.

In general, the patches look required, but I didn't run a full test against ORC 1.4.3.



Only ORC-296 might cause some performance difference.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon

**Comment by gatorsmile** [comment]
@dongjoon-hyun Could you add a test case for `ORC-285` in the Spark side? This sounds like a bug fix that impacts Spark?

**Comment by dongjoon-hyun** [comment]
I added a test case for ORC-285 and updated the JIRA and PR description.
Now, this PR aims to fix ORC-285 by updating ORC dependencies to 1.4.3.

**Comment by dongjoon-hyun** [comment]
This is `KafkaContinuousSourceStressForDontFailOnDataLossSuite` failure which
https://github.com/apache/spark/pull/20398 is trying to fix.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: This is the only issue for the new ORC reader, right? 

For the other two ORC reading modes, it works well, right? Could you also added test cases for the others too?

If this is the case, this should be a regression we have to merge to 2.3

File: pom.xml
Comment (gatorsmile) [inline pom.xml]: This is by ORC 1.4.2?

**Comment by gatorsmile** [comment]
I would still suggest to change the PR title back to `Upgrade ORC to 1.4.3`.

File: pom.xml
Comment (dongjoon-hyun) [reply pom.xml]: Yes. It's a regression introduced by ORC 1.4.1.
ORC 1.4.0 and 1.4.2, 1.4.3 is hiding this in a planned manner.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Yes. This is only issue for the new ORC reader, but I added this test case here for both `OrcSourceSuite` and `HiveOrcSourceSuite` to make it sure.

For Hive tables, if it needed, we had to create another test case. I'll try.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: yeah, for hive table, we might need a new test case

**Comment by dongjoon-hyun** [comment]
Title of PR/JIRA is updated, too.

File: pom.xml
Comment (gatorsmile) [reply pom.xml]: Thanks for the confirmation

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala
Comment (omalley) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: After a bit of digging, the problem that ORC-285 fixes was introduced in HIVE-9711. HIVE-9711 is included in Hive 1.2.0, but doesn't create a problem unless the vectorized ORC reader is used. That didn't become the default behavior until HIVE-11417, which was released in Hive 2.1.0. So Spark's other readers are fine as long as they don't use the vectorized reader.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Thank you, @omalley . Yes. It is. There is one difference in Spark.

Currently, Spark uses the vectorized reader for simple schema only. For `array<float>` or `array<double>` columns, we use ORC MR reader. However, MR reader uses vectorized RowBatch internally and it fails here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Since this is the issue when we use vectorized ORC reader, please check both vectorized and non-vectorized readers.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala]: This is not a test case for ORC-285

Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala]: Anyway, I removed it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala
Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: uh, I see. I just saw [your reply to omalley](https://github.com/apache/spark/pull/20511#discussion_r167952000)

**Comment by marmbrus** [comment]
Sorry if I'm missing some context here, but our typical process this late in the release (we are over a month since the branch was cut) would be to disable any new features that still have regressions.  I'm generally not okay with any dependency changes this far into QA unless there is no other option.

Should we be considering turning this off so that we can GA Spark 2.3 soon?

**Comment by gatorsmile** [comment]
Based on what @marmbrus suggested above, given how late it is in the cycle, we have to disable the native ORC reader (and also the filter pushdown) in 2.3.0 to avoid delaying the release, but add to release notes on how to turn them on manually. 

In the long term, our strategy is to improve our test coverage in the Spark side, instead of relying on the test cases of external data sources. 

cc @dongjoon-hyun @sameeragarwal @cloud-fan @rxin

**Comment by dongjoon-hyun** [comment]
+1.

**Comment by dongjoon-hyun** [comment]
@marmbrus , @sameeragarwal , @gatorsmile , @cloud-fan , @rxin .
Can we have this patch for Apache Spark 2.3.1?

**Comment by marmbrus** [comment]
Unfortunately, dependency changes are not typically allowed in patch releases.

**Comment by dongjoon-hyun** [comment]
I see. Thank you for confirmation, @marmbrus !

**Comment by gatorsmile** [comment]
Based on our standard, we do not change the dependencies in the minor releases.

**Comment by gatorsmile** [comment]
uh, I missed the previous comment.

**Comment by gatorsmile** [comment]
@dongjoon-hyun Could you rebase this PR? We want to merge it to master. Thanks!

**Comment by dongjoon-hyun** [comment]
Sure, I rebased this.

**Comment by dongjoon-hyun** [comment]
Thank you, All.
Now, it's ready for review again for Apache Spark 2.4.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Are we testing the native readers?

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Sure. This suite is in `sql/core` and inherits `OrcTest.scala`'s `val orcImp: String = "native"`.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Please also test both readers as we discussed above.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: ? I already added the test case at [HiveOrcQuerySuite.scala](https://github.com/apache/spark/pull/20511/files#diff-1569b2874975978ed62a01aab108d093R212), too.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: We have three ORC readers, right? We need to check all of them, and also vectorized reader too, even if they do not support it.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Ur, I think you are still confused two things.

First of all, we have five ORC readers. We didn't check `ORC MR reader` and `ORC Vectorized Copy` explicitly. We usually test 1, 2, and 3.

1. Hive Serde
2. Hive OrcFileFormat
3. Apache ORC Vectorized Wrapper
4. Apache ORC Vectorized Copy
5. Apache ORC MR

In this PR, we already adds 1, 2, 3. 3 is the vectorized wrapper reader.
1. Hive Serde : `HiveOrcQuerySuite.test(SPARK-23340 Empty float/double array columns raise EOFException)`
2. Hive OrcFileFormat: `OrcSourceSuite` <= `OrcSuite.test("SPARK-23340 Empty float/double array columns raise EOFException")`
3. Apache ORC Vectorized Wrapper: `HiveOrcSourceSuite` <= `OrcSuite.test("SPARK-23340 Empty float/double array columns raise EOFException")`

Second, this test schema includes complex types. So, 3 (vectorized wrapper reader) configuration is also going to fall-back ORC MR Reader path. In other words, case 5. Please note that Apache Spark support vectorization for `Atomic` types only in both Parquet and ORC.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: When we write these test cases, we should do our best to cover all the scenarios even if it is not supported now. This is like black-box testing.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: It's not about ORC issue. Spark doesn't allow complex type vectorization. Please see `def supportBatch`.
> When we write these test cases, we should do our best to cover all the scenarios even if it is not supported now. This is like black-box testing.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: You know, we still can support complex type vectorization in the future, but we might not re-visist all the tests in the future.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: Then, since the current case is already testing against vectorized configuration *by default*.

We can add `Seq(false, true).foreach(... spark.sql.orc.enableVectorizedReader <- value)`. we can add `vectorization=false` for now. However, both test cases will test MR reader only for now. If it sounds okay to you, then I'll add that.

**Comment by dongjoon-hyun** [comment]
@gatorsmile . 
In general, I agree with your idea to have a complete test coverage.
And, I think we can double the test coverage by enabling and disabling the vectorization for the whole ORC suites **instead of doing it in a specific test case**. How do you think about that? Can we do that kind of refactoring in a separate PR after this? If you want, I'll create Apache JIRA issue for that.

**Comment by gatorsmile** [comment]
@dongjoon-hyun That also sounds good to me. Always welcome to seeing more contributions to help improve the test coverage. Doing our best to improve the test coverage and code readability. 

Normally, when the reviewers have questions, (sometimes the questions are stupid), that might also mean the codes need more comments/refactoring. This can help the future code developers and maintainers.

**Comment by dongjoon-hyun** [comment]
Thank you. I created [SPARK-23452 Extend test coverage to all ORC readers](https://issues.apache.org/jira/browse/SPARK-23452) for that. Maybe, we can add more comments during that JIRA task.

**Comment by omalley** [comment]
I'm frustrated with the direction this has gone.

The new reader is much better than the old reader, which uses Hive 1.2. ORC 1.4.3 had a pair of important, but not large or complex fixes. Yet, because of those fixes, now the entire new reader is being disabled by default in the upcoming Spark 2.2.

In particular, the Hive 1.2 ORC code has the following known problems:
* HIVE-11312 - Char predicate pushdown can filter all rows
* HIVE-13083 - Decimal columns can incorrectly suppress the isNonNull stream.
* ORC-101 -  Predicate pushdown on bloom filters use default charset rather than UTF8.
* ORC-135 - Predicate pushdown on timestamps doesn't correct for timezones

**Comment by gatorsmile** [comment]
I agree on what @omalley said. The new reader based on ORC 1.4 is better than the old reader. That is why we chose the new reader as the default at the beginning. We also saw the performance improvement in the micro-benchmark. 

However, at the RC3 of Spark 2.3, we realize the new reader introduces the regression. After RC3, we reverted many PRs that caused the regressions and also rejected many bug fixes that could introduce new regressions. We are very conservative when merging the bug fixes in this stage. Thus, I also think the suggestion from @marmbrus is very reasonable. That is the strategy we are following in the previous Spark releases. 

Regarding this specific case, we did not revert the new reader, but just changed the default values. Users can try the new reader. We just want to avoid breaking the existing workloads when they upgrade to the upcoming Spark 2.3 release. In the next Spark 2.4 release, I believe we can feel more confident to choose the ORC new reader as the default. 

@dongjoon-hyun Could you submit a PR against the master branch to turn on them by default? Also add a migration guide.

**Comment by dongjoon-hyun** [comment]
Sure. @gatorsmile . I'll create a new PR for Apache Spark 2.4 default configuration and a migration guide for 2.3 to 2.4 after this PR is merged.

**Comment by gatorsmile** [comment]
LGTM 

Thanks! Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile !

**Pull request closed without merging**
